src/Workspaces/Core/Portable/EmbeddedLanguages/DateTime/LanguageServices/DateTimePatternDetector.cs

```diff
@@ -0,0 +1,202 @@
+                semanticModel, _ => TryCreate(semanticModel, info));
+        }
+
+        private static DateTimePatternDetector? TryCreate(
+            SemanticModel semanticModel, EmbeddedLanguageInfo info)
+        {
+            var dateTimeType = semanticModel.Compilation.GetTypeByMetadataName(typeof(System.DateTime).FullName!);
+            var dateTimeOffsetType = semanticModel.Compilation.GetTypeByMetadataName(typeof(System.DateTimeOffset).FullName!);
+            if (dateTimeType == null || dateTimeOffsetType == null)
+                return null;
+
+            return new DateTimePatternDetector(semanticModel, info, dateTimeType, dateTimeOffsetType);
+        }
+
+        public static bool IsDefinitelyNotDateTimeStringToken(SyntaxToken token, ISyntaxFacts syntaxFacts)
+        {
+            if (!syntaxFacts.IsStringLiteral(token))
+                return true;
+
+            if (!IsMethodArgument(token, syntaxFacts))
+                return true;
+
+            return false;
+        }
+
+        private static bool IsMethodArgument(SyntaxToken token, ISyntaxFacts syntaxFacts)
+            => syntaxFacts.IsLiteralExpression(token.Parent) &&
+               syntaxFacts.IsArgument(token.Parent!.Parent);
+
+        public bool IsDateTimeStringToken(SyntaxToken token, CancellationToken cancellationToken)
+        {
+            if (IsDefinitelyNotDateTimeStringToken(token, _info.SyntaxFacts))
+                return false;
+
+            var syntaxFacts = _info.SyntaxFacts;
+
+            var stringLiteral = token;
+            var literalNode = stringLiteral.Parent;
+            var argumentNode = literalNode!.Parent;
+            Debug.Assert(syntaxFacts.IsArgument(argumentNode));
+
+            var argumentList = argumentNode!.Parent;
+            var invocationOrCreation = argumentList!.Parent;
+            if (!syntaxFacts.IsInvocationExpression(invocationOrCreation))
+                return false;
+
+            var invokedExpression = syntaxFacts.GetExpressionOfInvocationExpression(invocationOrCreation);
+            var name = GetNameOfInvokedExpression(invokedExpression);
+            if (name != nameof(ToString) && name != nameof(System.DateTime.ParseExact) && name != nameof(System.DateTime.TryParseExact))
+                return false;
+
+            // if we couldn't determine the arg name or arg index, can't proceed.
+            var (argName, argIndex) = GetArgumentNameOrIndex(argumentNode);
+            if (argName == null && argIndex == null)
+                return false;
+
+            // If we had a specified arg name and it isn't 'format', then it's not a DateTime
+            // 'format' param we care about.
+            if (argName != null && argName != FormatName)
+                return false;
+
+            var symbolInfo = _semanticModel.GetSymbolInfo(invocationOrCreation, cancellationToken);
+            var method = symbolInfo.Symbol;
+            if (TryAnalyzeInvocation(method, argName, argIndex))
```

Comment:
think it would be better if we just used FormatName instead of passing in argName, since we already checked that argName must be FormatName above.

Reply:
sure!

Reply:
ah, that's not quite waht's going on here.  argName could be null.  We're effectively passing along tthe named arg if we have it.  or the argIndex otherwise.